### PR TITLE
SNOW-1707707: Add support for Index.to_numpy

### PR DESCRIPTION
**Pull request by sfc-gh-helmeleegy**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1707707

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [ ] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   Add support for Index.to_numpy.

### CODE REVIEW

File: src/snowflake/snowpark/modin/plugin/docstrings/series.py [hunk 1]
```diff
@@ -3383,6 +3383,61 @@ def to_numpy():
...
+        Index.array
+            Get the actual data stored within.
+        DataFrame.to_numpy
+            Similar method for DataFrame.
+
+        Notes
+        -----
+        The returned array will be the same up to equality (values equal in self will be equal in the returned array; likewise for values that are not equal). When self contains an ExtensionArray, the dtype may be different. For example, for a category-dtype Series, to_numpy() will return a NumPy array and the categorical dtype will be lost.
+
+        For NumPy dtypes, this will be a reference to the actual data stored in this Series or Index (assuming copy=False). Modifying the result in place will modify the data stored in the Series or Index (not that we recommend doing that).
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/series.py:3400]:
```suggestion
```

This part isn't true for Snowpark pandas.

File: src/snowflake/snowpark/modin/plugin/docstrings/series.py [hunk 2]
```diff
@@ -3383,6 +3383,61 @@ def to_numpy():
...
+
+        For extension types, to_numpy() may require copying data and coercing the result to a NumPy type (possibly object), which may be expensive. When you need a no-copy reference to the underlying data, Series.array should be used instead.
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/docstrings/series.py:3402]:
```suggestion
```

We never really copy the data, so I don't think this sentence makes sense either.

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 3]
```diff
@@ -2079,6 +2080,107 @@ def to_frame(
...
+        DataFrame.to_numpy
+            Similar method for DataFrame.
+
+        Notes
+        -----
+        The returned array will be the same up to equality (values equal in self will be equal in the returned array; likewise for values that are not equal). When self contains an ExtensionArray, the dtype may be different. For example, for a category-dtype Series, to_numpy() will return a NumPy array and the categorical dtype will be lost.
+
+        For NumPy dtypes, this will be a reference to the actual data stored in this Series or Index (assuming copy=False). Modifying the result in place will modify the data stored in the Series or Index (not that we recommend doing that).
+
+        For extension types, to_numpy() may require copying data and coercing the result to a NumPy type (possibly object), which may be expensive. When you need a no-copy reference to the underlying data, Series.array should be used instead.
```
Comment (sfc-gh-mvashishtha) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:2127]:
```suggestion
```

ditto about copy not being relevant for Snowpark pandas

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 4]
```diff
@@ -2079,6 +2080,107 @@ def to_frame(
...
+        array(['1999-12-31T23:00:00.000000000', '2000-01-01T23:00:00...'],
+            dtype='datetime64[ns]')
+        """
+        if copy:
+            WarningMessage.ignored_argument(
+                operation="to_numpy",
+                argument="copy",
+                message="copy is ignored in Snowflake backend",
+            )
+        # return self.to_pandas().array
```
Comment (sfc-gh-jjiao) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:2173]: could we delete this commented line?

File: src/snowflake/snowpark/modin/plugin/extensions/index.py [hunk 5]
```diff
@@ -2601,6 +2698,10 @@ def __array__(self, dtype: Any = None) -> np.ndarray:
         """
         The array interface, return the values.
         """
+        # Ensure that the existing index dtype is preserved in the returned array
+        # if no other dtype is given.
+        if dtype is None:
```
Comment (sfc-gh-helmeleegy) [inline src/snowflake/snowpark/modin/plugin/extensions/index.py:2703]: This fixes a pre-existing issue that was causing some test failures with the introduction of `Index.to_numpy`.

File: tests/integ/modin/index/test_astype.py [hunk 6]
```diff
@@ -166,8 +168,8 @@ def test_index_astype_bool_nan_none():
     snow_index = pd.Index(native_index)
     with pytest.raises(AssertionError):
         assert_index_equal(snow_index.astype(bool), native_index.astype(bool))
-    expected_result = native_pd.Index([True, True, True, False, False], dtype=bool)
-    assert_index_equal(snow_index.astype(bool), expected_result)
+    expected_result = native_pd.Index([True, True, True, None, None], dtype=bool)
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/index/test_astype.py:171]: This is more consistent with pandas now.

File: tests/integ/modin/frame/test_getitem.py [hunk 7]
```diff
@@ -36,31 +36,28 @@
...
+@sql_count_checker(query_count=1, join_count=1)
 def test_df_getitem_with_boolean_list_like(
     key, default_index_snowpark_pandas_df, default_index_native_df
 ):
-    # one added query to convert to native pandas and 1 added query for series initialization
-    with SqlCounter(
-        query_count=3 if isinstance(key, native_pd.Index) else 1, join_count=1
-    ):
-        # df[boolean list-like key] is the same as df.loc[:, boolean list-like key]
-        if isinstance(key, native_pd.Index):
```
Comment (sfc-gh-helmeleegy) [inline tests/integ/modin/frame/test_getitem.py:47]: There was no need to convert the native pandas index into a Snowpark pandas index when the df it's being passed to is actually a native pandas DataFrame. This was also causing test failures on the introduction of `Index.to_numpy`

**Review by sfc-gh-mvashishtha — APPROVED** [review]
thanks!

**Pull request merged**